File: MoonEngine-Root/Assets/src/main/java/com/greentree/engine/moon/assets/serializator/AbstractIterableAssetSerializator.java
package com.greentree.engine.moon.assets.serializator;

import com.greentree.commons.util.classes.info.TypeInfo;
import com.greentree.commons.util.classes.info.TypeInfoBuilder;
import com.greentree.commons.util.classes.info.TypeUtil;

public abstract class AbstractIterableAssetSerializator<T>
		implements AssetSerializator<Iterable<T>> {
	
	public final TypeInfo<T> ITER_TYPE;
	
	public final TypeInfo<Iterable<T>> TYPE;
	
	public AbstractIterableAssetSerializator(Class<T> cls) {
		this(TypeInfoBuilder.getTypeInfo(cls));
	}
	
	public AbstractIterableAssetSerializator(TypeInfo<T> type) {
		ITER_TYPE = type;
		TYPE = TypeInfoBuilder.getTypeInfo(Iterable.class, ITER_TYPE);
	}
	
	protected AbstractIterableAssetSerializator() {
		ITER_TYPE = TypeUtil.getFirstAtgument(TypeInfoBuilder.getTypeInfo(getClass()),
				AbstractIterableAssetSerializator.class);
		TYPE = TypeInfoBuilder.getTypeInfo(Iterable.class, ITER_TYPE);
	}
	
	@Override
	public final TypeInfo<Iterable<T>> getType() {
		return TYPE;
	}
	
}

File: MoonEngine-Root/Assets/src/main/java/com/greentree/engine/moon/assets/serializator/AbstractIterableAssetSerializator.kt
package com.greentree.engine.moon.assets.serializator

import com.greentree.commons.reflection.info.TypeInfo
import com.greentree.commons.reflection.info.TypeInfoBuilder
import com.greentree.commons.reflection.info.TypeUtil

abstract class AbstractIterableAssetSerializator<T> : AssetSerializator<Iterable<T>> {

	val ITER_TYPE: TypeInfo<T>
	override val type: TypeInfo<Iterable<T>>

	constructor(cls: Class<T>?) : this(TypeInfoBuilder.getTypeInfo<T>(cls))
	constructor(type: TypeInfo<T>) {
		ITER_TYPE = type
		this.type = TypeInfoBuilder.getTypeInfo(Iterable::class.java, ITER_TYPE)
	}

	protected constructor() {
		ITER_TYPE = TypeUtil.getFirstAtgument(
			TypeInfoBuilder.getTypeInfo(javaClass),
			AbstractIterableAssetSerializator::class.java
		)
		type = TypeInfoBuilder.getTypeInfo(Iterable::class.java, ITER_TYPE)
	}
}